### PR TITLE
OID Generation

### PR DESCRIPTION
**Pull request by kyeah**
Provides an implementation for client-side generation of ObjectIds. Interfaces with `libc` are necessary for retrieving the machine and process ids, introducing a couple of `unsafe` blocks.

### CODE REVIEW

File: tests/client/oid.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    let s = "541b1a00e8a23afa832b218e";
+    let oid_res = oid::with_string(s);
+    assert!(oid_res.is_ok());
+    let oid = oid_res.unwrap();
+    let bytes: [u8; 12] = [0x54u8, 0x1Bu8, 0x1Au8, 0x00u8,
+                           0xE8u8, 0xA2u8, 0x3Au8, 0xFAu8,
+                           0x83u8, 0x2Bu8, 0x21u8, 0x8Eu8];
+
+    for i in 0..12 {
+        assert_eq!(bytes[i], oid[i]);
```
Comment (alabid) [inline tests/client/oid.rs]: `oid` should implement the `core::cmp::Eq` trait. There are many situations when a user would want to compare object ids.

File: tests/client/oid.rs [hunk 2]
```diff
@@ -0,0 +1,57 @@
+use mongodb::client::oid;
+use rustc_serialize::hex::ToHex;
+
```
Comment (alabid) [inline tests/client/oid.rs]: I want to see some test cases testing endianness. For example in the server, we have: https://github.com/mongodb/mongo/blob/master/src/mongo/bson/oid_test.cpp#L77

File: src/client/oid.rs [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+    let mut buf: [u8; 8] = [0; 8];
+    BigEndian::write_u64(&mut buf, u_int);
+
+    let buf_u24: [u8; 3] = [buf[0], buf[1], buf[2]];
+    Ok(buf_u24)
+}
+
+#[test]
+fn pid_generation() {
```
Comment (alabid) [inline src/client/oid.rs]: why is this test here and not in `tests/client/oid.rs`?

File: tests/client/oid.rs (hunk 1 above)
Comment (kyeah) [inline tests/client/oid.rs]: oops, this can actually be compared directly (including inequalities).

File: src/client/oid.rs [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    // Init oid counter
+    if OID_COUNTER.load(Ordering::SeqCst) == 0 {
+        let mut rng = try!(OsRng::new());
+        let start = rng.gen_range(0, MAX_U24 + 1);
+        OID_COUNTER.store(start, Ordering::SeqCst);
+    }
+
+    let u_counter = OID_COUNTER.fetch_add(1, Ordering::SeqCst);
+
+    // Mod result instead of OID_COUNTER to prevent threading issues.
```
Comment (alabid) [inline src/client/oid.rs]: cool

File: src/client/oid.rs (hunk 3 above)
Comment (kyeah) [inline src/client/oid.rs]: These are unit tests that rely on private functions, so they can't be isolated into a separate test module.

File: src/client/oid.rs [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+    let mut buf: [u8; 4] = [0; 4];
+    for i in 0..3 {
+        buf[i] = oid[4+i];
+    }
+    LittleEndian::read_u32(&buf)
+}
+
+/// Retrieves the process id associated with an ObjectId.
+pub fn get_pid(oid: [u8; 12]) -> u16 {
+    LittleEndian::read_u16(&oid[7..])
```
Comment (alabid) [inline src/client/oid.rs]: I don't want to see "magic" number around. Define special numbers as constant. For example: `kTimestampOffset`, `kInstanceUniqueOffset`, and `kIncrementOffset` instantiated here: thttps://github.com/mongodb/mongo/blob/master/src/mongo/bson/oid.cpp#L47

File: src/client/oid.rs [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+/// for more information.
+pub fn generate() -> Result<[u8; 12]> {
+    let timestamp = gen_timestamp();
+    let machine_id = try!(gen_machine_id());
+    let process_id = gen_pid();
+    let counter = try!(gen_count());
+
+    let mut buf: [u8; 12] = [0; 12];
+    for i in 0..4 { buf[i] = timestamp[i]; }
+    for i in 0..3 { buf[4+i] = machine_id[i]; }
```
Comment (alabid) [inline src/client/oid.rs]: Don't use magic numbers.

File: Cargo.lock [hunk 7]
```diff
@@ -4,8 +4,12 @@ version = "0.1.0"
 dependencies = [
  "bson 0.1.2 (git+https://github.com/kyeah/bson-rs)",
  "byteorder 0.3.10 (registry+https://github.com/rust-lang/crates.io-index)",
+ "libc 0.1.8 (registry+https://github.com/rust-lang/crates.io-index)",
```
Comment (alabid) [inline Cargo.lock]: I like the way the dependencies here are listed lexicographically. Maybe we should do the same when using crates? What does the rust style guide say about that?

File: src/client/oid.rs (hunk 3 above)
Comment (alabid) [inline src/client/oid.rs]: Ok.

File: Cargo.lock (hunk 7 above)
Comment (kyeah) [inline Cargo.lock]: The guide doesn't seem to say anything explicitly, but they should be listed lexicographically; looks like that's the case in Cargo.toml and lib.rs. Are you talking about module-level imports (`use ...`)?

Comment (alabid) [inline Cargo.lock]: yep.

Comment (saghmrossi) [inline Cargo.lock]: This is a generally good idea, but in a couple edges cases we'd have to not to that in order for things to work. The main one I'm thinking of is when a module with macros is imported in lib.rs/mod.rs; the macros will only be accessible for the modules declared *after* the import/declaration of the module with macros, so those ones would have to appear first.

**Pull request merged**